File: android/app/src/main/java/com/vegettable/app/ui/compare/CompareActivity.java
package com.vegettable.app.ui.compare;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.textfield.TextInputEditText;
import com.vegettable.app.R;
import com.vegettable.app.model.ApiResponse;
import com.vegettable.app.model.MarketPrice;
import com.vegettable.app.network.ApiClient;
import com.vegettable.app.util.PriceUtils;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class CompareActivity extends AppCompatActivity {

    private TextInputEditText etCropName;
    private RecyclerView rvCompare;
    private ProgressBar progressBar;
    private LinearLayout layoutError;
    private TextView tvError;
    private CompareAdapter adapter;
    private String lastCropName;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_compare);

        findViewById(R.id.btn_back).setOnClickListener(v -> finish());

        etCropName = findViewById(R.id.et_crop_name);
        rvCompare = findViewById(R.id.rv_compare);
        progressBar = findViewById(R.id.progress_bar);
        layoutError = findViewById(R.id.layout_error);
        tvError = findViewById(R.id.tv_error);

        rvCompare.setLayoutManager(new LinearLayoutManager(this));
        adapter = new CompareAdapter();
        rvCompare.setAdapter(adapter);

        findViewById(R.id.btn_retry).setOnClickListener(v -> {
            if (lastCropName != null && !lastCropName.isEmpty()) {
                comparePrices(lastCropName);
            }
        });

        etCropName.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                String crop = etCropName.getText().toString().trim();
                if (!crop.isEmpty()) {
                    comparePrices(crop);
                }
                return true;
            }
            return false;
        });
    }

    private void comparePrices(String cropName) {
        lastCropName = cropName;
        progressBar.setVisibility(View.VISIBLE);
        layoutError.setVisibility(View.GONE);

        ApiClient.getInstance().getApi().compareMarketPrices(cropName, null)
                .enqueue(new Callback<ApiResponse<List<MarketPrice>>>() {
                    @Override
                    public void onResponse(@NonNull Call<ApiResponse<List<MarketPrice>>> call,
                                           @NonNull Response<ApiResponse<List<MarketPrice>>> response) {
                        progressBar.setVisibility(View.GONE);
                        if (response.isSuccessful() && response.body() != null
                                && response.body().isSuccess() && response.body().getData() != null) {
                            adapter.setItems(response.body().getData());
                            rvCompare.setVisibility(View.VISIBLE);
                        } else {
                            showError("無法取得比價資料");
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<ApiResponse<List<MarketPrice>>> call,
                                          @NonNull Throwable t) {
                        progressBar.setVisibility(View.GONE);
                        showError("網路錯誤: " + t.getMessage());
                    }
                });
    }

    private void showError(String message) {
        tvError.setText(message);
        layoutError.setVisibility(View.VISIBLE);
    }

    // ─── Adapter ────────────────────────────────────────────
    static class CompareAdapter extends RecyclerView.Adapter<CompareAdapter.VH> {
        private List<MarketPrice> items = new ArrayList<>();

        void setItems(List<MarketPrice> items) {
            this.items = items;
            notifyDataSetChanged();
        }

        @NonNull @Override
        public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View v = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_compare, parent, false);
            return new VH(v);
        }

        @Override
        public void onBindViewHolder(@NonNull VH h, int pos) {
            MarketPrice mp = items.get(pos);
            h.tvMarketName.setText(mp.getMarketName());
            h.tvDate.setText(mp.getTransDate());
            h.tvAvgPrice.setText(PriceUtils.formatPrice(mp.getAvgPrice()) + " 元/公斤");
            h.tvPriceRange.setText(PriceUtils.formatPrice(mp.getLowerPrice())
                    + " ~ " + PriceUtils.formatPrice(mp.getUpperPrice()));
        }

        @Override public int getItemCount() { return items.size(); }

        static class VH extends RecyclerView.ViewHolder {
            TextView tvMarketName, tvDate, tvAvgPrice, tvPriceRange;
            VH(View v) {
                super(v);
                tvMarketName = v.findViewById(R.id.tv_market_name);
                tvDate = v.findViewById(R.id.tv_date);
                tvAvgPrice = v.findViewById(R.id.tv_avg_price);
                tvPriceRange = v.findViewById(R.id.tv_price_range);
            }
        }
    }
}

File: android/app/src/main/java/com/vegettable/app/ui/compare/CompareActivity.kt
package com.vegettable.app.ui.compare

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.textfield.TextInputEditText
import com.vegettable.app.R
import com.vegettable.app.model.ApiResponse
import com.vegettable.app.model.MarketPrice
import com.vegettable.app.network.ApiClient.Companion.instance
import com.vegettable.app.util.PriceUtils
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class CompareActivity : AppCompatActivity() {
    private var etCropName: TextInputEditText? = null
    private var rvCompare: RecyclerView? = null
    private var progressBar: ProgressBar? = null
    private var layoutError: LinearLayout? = null
    private var tvError: TextView? = null
    private var adapter: CompareAdapter? = null
    private var lastCropName: String? = null
    private var retryCount = 0
    private val retryHandler = Handler(Looper.getMainLooper())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_compare)

        findViewById<View?>(R.id.btn_back).setOnClickListener { finish() }

        etCropName = findViewById(R.id.et_crop_name)
        rvCompare = findViewById(R.id.rv_compare)
        progressBar = findViewById(R.id.progress_bar)
        layoutError = findViewById(R.id.layout_error)
        tvError = findViewById(R.id.tv_error)

        rvCompare?.layoutManager = LinearLayoutManager(this)
        adapter = CompareAdapter()
        rvCompare?.adapter = adapter

        findViewById<View?>(R.id.btn_retry).setOnClickListener {
            if (!lastCropName.isNullOrEmpty()) {
                val delay = when (retryCount) {
                    0 -> 0L
                    1 -> 2000L
                    2 -> 4000L
                    else -> 8000L
                }
                retryCount++
                retryHandler.postDelayed({ comparePrices(lastCropName) }, delay)
            }
        }

        etCropName?.setOnEditorActionListener { v, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                val crop = etCropName?.text.toString().trim()
                if (crop.isNotEmpty()) {
                    val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
                    imm.hideSoftInputFromWindow(v.windowToken, 0)
                    comparePrices(crop)
                }
                true
            } else {
                false
            }
        }
    }

    private fun comparePrices(cropName: String?) {
        lastCropName = cropName
        progressBar?.visibility = View.VISIBLE
        layoutError?.visibility = View.GONE
        rvCompare?.visibility = View.GONE

        instance?.api?.compareMarketPrices(cropName, null)?.enqueue(object : Callback<ApiResponse<MutableList<MarketPrice?>?>?> {
            override fun onResponse(
                call: Call<ApiResponse<MutableList<MarketPrice?>?>?>,
                response: Response<ApiResponse<MutableList<MarketPrice?>?>?>
            ) {
                progressBar?.visibility = View.GONE
                val body = response.body()
                if (response.isSuccessful && body != null && body.isSuccess && body.data != null) {
                    val marketPrices = body.data?.filterNotNull()?.toMutableList() ?: mutableListOf()
                    adapter?.setItems(marketPrices)
                    retryCount = 0
                    if (marketPrices.isEmpty()) {
                        showError("找不到「${lastCropName}」的市場比價資料")
                    } else {
                        rvCompare?.visibility = View.VISIBLE
                    }
                } else {
                    showError("無法取得比價資料")
                }
            }

            override fun onFailure(
                call: Call<ApiResponse<MutableList<MarketPrice?>?>?>,
                t: Throwable
            ) {
                progressBar?.visibility = View.GONE
                showError("網路錯誤: " + t.message)
            }
        })
    }

    private fun showError(message: String?) {
        tvError?.text = message
        layoutError?.visibility = View.VISIBLE
    }

    // ─── Adapter ────────────────────────────────────────────
    internal class CompareAdapter : RecyclerView.Adapter<CompareAdapter.VH>() {
        private var items: MutableList<MarketPrice> = ArrayList()

        fun setItems(items: MutableList<MarketPrice>) {
            this.items = items
            notifyDataSetChanged()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
            val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_compare, parent, false)
            return VH(v)
        }

        override fun onBindViewHolder(h: VH, pos: Int) {
            val mp = items[pos]
            h.tvMarketName.text = mp.marketName
            h.tvDate.text = mp.transDate
            h.tvAvgPrice.text = PriceUtils.formatPrice(mp.avgPrice) + " 元/公斤"
            h.tvPriceRange.text = (PriceUtils.formatPrice(mp.lowerPrice) + " ~ " + PriceUtils.formatPrice(mp.upperPrice))
        }

        override fun getItemCount(): Int {
            return items.size
        }

        internal class VH(v: View) : RecyclerView.ViewHolder(v) {
            var tvMarketName: TextView = v.findViewById(R.id.tv_market_name)
            var tvDate: TextView = v.findViewById(R.id.tv_date)
            var tvAvgPrice: TextView = v.findViewById(R.id.tv_avg_price)
            var tvPriceRange: TextView = v.findViewById(R.id.tv_price_range)
        }
    }
}
